posts/2018-09-04-chain-replacing-the-list-monoid.md

```diff
@@ -0,0 +1,66 @@
+---
+layout: post
+title: Chain – Replacing the List Monoid
+
+meta:
+  nav: blog
+  author: lukajcb
+  pygments: true
+
+---
+
+`List` is a great data type, it is very simple and easy to understand.
+It has very low overhead for the most important functions such as `fold` and `map` and also supports prepending a single element in constant time.
+
+Traversing a data structure with something like `Writer[List[Log], A]` or `ValidatedNel[Error, A]`, is really powerful and allows us to very precisely specify what kind of iteration we want to do while remaining very succint.
+However, in terms of efficiency it's a whole different story unfortunately.
+That is because both of these traversals make use of the `List` monoid (or the `NonEmptyList` semigroup), which by the nature of `List` is very inefficient.
+If you use `traverse` with a data structure with `n` elements and `Writer` or `Validated` as the `Applicative` type, you will end up with a runtime of `O(n^2)`.
+This is because, with `List`, appending a single element requires iterating over the entire data structure and therefore takes linear time.
+
+So `List` isn't all that great for this use case, so let's use `Vector` or `NonEmptyVector` instead, right?
+
+Well, `Vector` has its own problems and in this case it's unfortunately not all that faster than `List` at all. You can check [this blog post](http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html#vectors-are-ok) by Li Haoyi for some deeper insight into `Vector`s issues.
+
+Because of this, it's now time to welcome a new data structure to Cats.
+Meet `Chain` and it's non-empty counterpart `NonEmptyChain`. 
+
+Available in the newest Cats 1.3.0 release, `Chain` is somewhat derived from what used to be `fs2.Catenable` and Erik Osheim's [Chain](https://github.com/non/chain ) library.
+Similar to `List` it is also a very simple data structure, but unlike `List` it supports both constant time `append` and `prepend`.
```

Comment:
> guarantees constant time for both `append` and `prepend`?

Always unsure on wheter to prefer `O(1)` or `constant`

Reply:
I'll just add both :P